From 0cd84b7bdaf51facf6689c135895911b4d8dc758 Mon Sep 17 00:00:00 2001
From: Ken Ma <make@marvell.com>
Date: Fri, 23 Jun 2017 09:47:38 +0800
Subject: [PATCH 0205/1239] net: switch: mv88e6xxx: add port setup in
 initialization

With default setting in port control register, switch works well; but
linux kernel will change the port control setting, for example, it will
add tag for egress frames in cpu port; then after reboot in kernel,
the changed port control setting is kept in u-boot, but the changed
setting does not work for switch in u-boot.
This patch adds port setup in initialization to make sure that switch
can work properly in u-boot.

Change-Id: Ic380f80f712707003668172d6092d1603f54a570
Signed-off-by: Ken Ma <make@marvell.com>
Signed-off-by: Igal Liberman <igall@marvell.com>
---
 drivers/net/phy/mv88e6xxx.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/drivers/net/phy/mv88e6xxx.c b/drivers/net/phy/mv88e6xxx.c
index 663b1a0ec9..606ded5132 100644
--- a/drivers/net/phy/mv88e6xxx.c
+++ b/drivers/net/phy/mv88e6xxx.c
@@ -338,6 +338,7 @@ int mv88e6xxx_initialize(const void *blob)
 {
 	int node = 0;
 	int ret;
+	int port;
 
 	soho_dev_handle = NULL;
 
@@ -389,6 +390,28 @@ int mv88e6xxx_initialize(const void *blob)
 			return ret;
 	}
 
+	/* Force port setup */
+	for (port = 0; port < sizeof(soho_dev.port_mask) * 8; port++) {
+		if (!(soho_dev.port_mask & BIT(port)))
+			continue;
+
+		/* Set port control register */
+		mv88e6xxx_write_register(&soho_dev,
+					 REG_PORT(port),
+					 PORT_CONTROL,
+					 PORT_CONTROL_STATE_FORWARDING |
+					 PORT_CONTROL_FORWARD_UNKNOWN |
+					 PORT_CONTROL_FORWARD_UNKNOWN_MC |
+					 PORT_CONTROL_USE_TAG |
+					 PORT_CONTROL_USE_IP |
+					 PORT_CONTROL_TAG_IF_BOTH);
+		/* Set port based vlan table */
+		mv88e6xxx_write_register(&soho_dev,
+					 REG_PORT(port),
+					 PORT_BASE_VLAN,
+					 soho_dev.port_mask & ~BIT(port));
+	}
+
 	soho_dev_handle = &soho_dev;
 
 	return 0;
-- 
2.29.0

